Applications/Spire/Source/Ui/Tooltip.cpp

```diff
@@ -0,0 +1,237 @@
+  // TODO: Add drop shadow if this gets fixed:
+  //       https://bugreports.qt.io/browse/QTBUG-58602
+  //auto shadow = new QGraphicsDropShadowEffect(this);
+  //shadow->setColor(DROP_SHADOW_COLOR);
+  //shadow->setOffset(0, 0);
+  //shadow->setBlurRadius(scale_width(5));
+  //setGraphicsEffect(shadow);
+}
+
+bool Tooltip::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_body) {
+    if(event->type() == QEvent::Resize) {
+      resize(m_body->size().grownBy(get_margins()));
+    }
+  } else if(watched == parentWidget()) {
+    switch(event->type()) {
+      case QEvent::HoverEnter:
+        if(parentWidget()->isEnabled()) {
+          m_show_timer.start();
+        }
+        break;
+      case QEvent::MouseMove:
+        if(!parentWidget()->rect().contains(
+            static_cast<QMouseEvent*>(event)->pos())) {
+          hide();
+        }
+        break;
+      case QEvent::HoverLeave:
+      case QEvent::WindowDeactivate:
+        hide();
+        break;
+      case QEvent::ToolTip:
+        return true;
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void Tooltip::hideEvent(QHideEvent* event) {
+  m_show_timer.stop();
+  QWidget::hideEvent(event);
+}
+
+void Tooltip::paintEvent(QPaintEvent* event) {
+  auto painter = QPainter(this);
+  painter.fillRect(rect().marginsRemoved(get_margins()), TOOLTIP_COLOR);
+  painter.fillPath(get_arrow_path(), TOOLTIP_COLOR);
+}
+
+QPainterPath Tooltip::get_arrow_path() const {
+  auto path = QPainterPath();
+  auto polygon = [&] () -> QPolygonF {
+    auto margins = get_margins();
+    auto left_x = ARROW_X_POSITION() + DROP_SHADOW_WIDTH();
+    auto tip_x = left_x + (ARROW_SIZE().width() / 2);
+    auto right_x = left_x + ARROW_SIZE().width();
+    auto orientation = get_orientation();
+    if(orientation == Orientation::TOP_LEFT ||
+        orientation == Orientation::TOP_RIGHT) {
+      return QVector<QPoint>({{left_x, height() - margins.bottom()},
+        {tip_x, height() - Y_OFFSET()},
+        {right_x, height() - margins.bottom()}});
+    }
+    return QVector<QPoint>({{left_x, margins.top()}, {tip_x, Y_OFFSET()},
+      {right_x, margins.top()}});
+  }();
+  if(get_body_orientation() == BodyOrientation::LEFT) {
+    polygon.translate(width() - (2 * (ARROW_X_POSITION() +
+      DROP_SHADOW_WIDTH())) - ARROW_SIZE().width(), 0);
+  }
+  path.addPolygon(polygon);
+  return path;
+}
+
+Tooltip::BodyOrientation Tooltip::get_body_orientation() const {
+  auto parent_position = parentWidget()->mapToGlobal(
+    parentWidget()->rect().bottomLeft());
+  auto screen_geometry =
+    get_current_screen(parent_position)->availableGeometry();
+  if(parent_position.x() + width() >
+      screen_geometry.x() + screen_geometry.width()) {
+    return BodyOrientation::LEFT;
+  }
+  return BodyOrientation::RIGHT;
+}
+
+QScreen* Tooltip::get_current_screen(const QPoint& point) const {
+  if(auto screen = QGuiApplication::screenAt(point); screen != nullptr) {
```

Comment:
Remove the ```; screen != nullptr```.

Reply:
So declare screen before the if statement?

Reply:
Nope, refer to:

https://en.cppreference.com/w/cpp/language/if

The definition of condition:
 declaration of a single non-array variable with a brace-or-equals initializer.

Reply:
Okay, I see. I remember there was one case where I used the semicolon, but that must have been because the variable couldn't be evaluate as a boolean.